## Final Project Submission

Please fill out:
* Student name: Yvonnie Muthoni Wanyoike
* Student pace: full time
* Scheduled project review date/time: 19/12/2025
* Instructor name: Maryann Mwikali
* Blog post URL:


# Aircraft Risk Assessment Project

## Project Overview

This project identifies the lowest-risk aircraft for the company’s new aviation division by analyzing historical flight safety data.The aviation accident historical data is from the National Transportation Safety Board from 1962 to 2023 and its about civil aviation accidents and selected incidents in the United States and international waters.Through thorough data cleaning, imputation, and risk assessment, I transformed raw accident records into actionable insights to guide strategic purchasing decisions and minimize operational liability.

## Business Problem

Our organization is entering the aviation sector with no prior operational history, creating significant exposure to safety and financial liabilities. Without a data-driven understanding of aircraft reliability, the company risks investing in high-maintenance or accident-prone models. This project identifies aircrafts with the highest safety ratings and lowest historical risk profiles to ensure a secure market entry.

## Objectives

1. To develop a list of aircraft makes, models and categories categorized by their safety performance.
2. To identify the aircrafts that demonstrate the highest-operational resilience-those that remain safe even in challenging environments or critical flight phases.


## Accessing Data

### Importing Data

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

### Loading Data

Use pandas to create a new DataFrame, called `df`, containing the data from the dataset in the file `Aviation_Data.csv` in the folder containing this notebook. 


In [2]:
df = pd.read_csv('Aviation_Data.csv',low_memory=False)
df

Event.Id Investigation.Type Accident.Number  Event.Date  \
0      20001218X45444           Accident      SEA87LA080  1948-10-24   
1      20001218X45447           Accident      LAX94LA336  1962-07-19   
2      20061025X01555           Accident      NYC07LA005  1974-08-30   
3      20001218X45448           Accident      LAX96LA321  1977-06-19   
4      20041105X01764           Accident      CHI79FA064  1979-08-02   
...               ...                ...             ...         ...   
90343  20221227106491           Accident      ERA23LA093  2022-12-26   
90344  20221227106494           Accident      ERA23LA095  2022-12-26   
90345  20221227106497           Accident      WPR23LA075  2022-12-26   
90346  20221227106498           Accident      WPR23LA076  2022-12-26   
90347  20221230106513           Accident      ERA23LA097  2022-12-29   

              Location        Country   Latitude   Longitude Airport.Code  \
0      MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1       BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2        Saltville, VA  United States  36.922223  -81.878056          NaN   
3           EUREKA, CA  United States        NaN         NaN          NaN   
4           Canton, OH  United States        NaN         NaN          NaN   
...                ...            ...        ...         ...          ...   
90343    Annapolis, MD  United States        NaN         NaN          NaN   
90344      Hampton, NH  United States        NaN         NaN          NaN   
90345       Payson, AZ  United States    341525N    1112021W          PAN   
90346       Morgan, UT  United States        NaN         NaN          NaN   
90347       Athens, GA  United States        NaN         NaN          NaN   

      Airport.Name  ... Purpose.of.flight         Air.carrier  \
0              NaN  ...          Personal                 NaN   
1              NaN  ...          Personal                 NaN   
2              NaN  ...          Personal                 NaN   
3              NaN  ...          Personal                 NaN   
4              NaN  ...          Personal                 NaN   
...            ...  ...               ...                 ...   
90343          NaN  ...          Personal                 NaN   
90344          NaN  ...               NaN                 NaN   
90345       PAYSON  ...          Personal                 NaN   
90346          NaN  ...          Personal  MC CESSNA 210N LLC   
90347          NaN  ...          Personal                 NaN   

      Total.Fatal.Injuries Total.Serious.Injuries Total.Minor.Injuries  \
0                      2.0                    0.0                  0.0   
1                      4.0                    0.0                  0.0   
2                      3.0                    NaN                  NaN   
3                      2.0                    0.0                  0.0   
4                      1.0                    2.0                  NaN   
...                    ...                    ...                  ...   
90343                  0.0                    1.0                  0.0   
90344                  0.0                    0.0                  0.0   
90345                  0.0                    0.0                  0.0   
90346                  0.0                    0.0                  0.0   
90347                  0.0                    1.0                  0.0   

      Total.Uninjured Weather.Condition  Broad.phase.of.flight  \
0                 0.0               UNK                 Cruise   
1                 0.0               UNK                Unknown   
2                 NaN               IMC                 Cruise   
3                 0.0               IMC                 Cruise   
4                 0.0               VMC               Approach   
...               ...               ...                    ...   
90343             0.0               NaN                    NaN   
90344             0.0    

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90348 entries, 0 to 90347
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      90348 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            50249 non-null  object 
 9   Airport.Name            52790 non-null  object 
 10  Injury.Severity         87889 non-null  object 
 11  Aircraft.damage         85695 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     87572 non-null  object 
 14  Make                    88826 non-null

#### Interpreting the Results
1. There are 90348 rows and 31 columns in the dataframe.
2. 5 columns contain numeric data while 26 contain categorical data.
3. Most of the columns contain missing values.

In [4]:
#checking the columns in the dataframe
df.columns

Index(['Event.Id', 'Investigation.Type', 'Accident.Number', 'Event.Date',
       'Location', 'Country', 'Latitude', 'Longitude', 'Airport.Code',
       'Airport.Name', 'Injury.Severity', 'Aircraft.damage',
       'Aircraft.Category', 'Registration.Number', 'Make', 'Model',
       'Amateur.Built', 'Number.of.Engines', 'Engine.Type', 'FAR.Description',
       'Schedule', 'Purpose.of.flight', 'Air.carrier', 'Total.Fatal.Injuries',
       'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured',
       'Weather.Condition', 'Broad.phase.of.flight', 'Report.Status',
       'Publication.Date'],
      dtype='object')

## Data Cleaning

### Handling missing values

I'm checking for missing values in important columns(**working cols**) which are: Aircraft.Category,Registration.Number,Make,Model,Amateur.Built, No.of.Engines,Engine.Type,Schedule,Purpose.of.flight,Total.Fatal.Injuries,Total.Serious.Injuries,Total.Minor.Injuries,Total.Uninjured,Weather.Conditionand Broad.phase.of.flight.Quite a number of them but I believe all these are factors in this analysis.

In [5]:
working_cols = ['Aircraft.Category','Registration.Number','Make','Model','Amateur.Built','Number.of.Engines'
                      ,'Engine.Type','Schedule','Purpose.of.flight','Total.Fatal.Injuries','Total.Serious.Injuries'
                      ,'Total.Minor.Injuries','Total.Uninjured','Weather.Condition','Broad.phase.of.flight']

In [6]:
#checking missing values in my working columns
df[working_cols].isna().sum()

Aircraft.Category         58061
Registration.Number        2776
Make                       1522
Model                      1551
Amateur.Built              1561
Number.of.Engines          7543
Engine.Type                8536
Schedule                  77766
Purpose.of.flight          7651
Total.Fatal.Injuries      12860
Total.Serious.Injuries    13969
Total.Minor.Injuries      13392
Total.Uninjured            7371
Weather.Condition          5951
Broad.phase.of.flight     28624
dtype: int64

In [7]:
#checking the missing values in % form to estimate their relevance to the dataset
df[working_cols].isna().mean()*100

Aircraft.Category         64.263736
Registration.Number        3.072564
Make                       1.684597
Model                      1.716695
Amateur.Built              1.727764
Number.of.Engines          8.348829
Engine.Type                9.447913
Schedule                  86.073848
Purpose.of.flight          8.468367
Total.Fatal.Injuries      14.233851
Total.Serious.Injuries    15.461327
Total.Minor.Injuries      14.822686
Total.Uninjured            8.158454
Weather.Condition          6.586753
Broad.phase.of.flight     31.681941
dtype: float64

 According to the above output,I'll have to drop the rows with the missing values `Nan` in 'Registration.Number' column because the Registration number is a unique identifier therefore I can't replace the missing values with other values. 

In [8]:
#identifying the unique registration numbers
df['Registration.Number'].unique()

array(['NC6404', 'N5069P', 'N5142R', ..., 'N749PJ', 'N210CU', 'N9026P'],
      dtype=object)

In [9]:
#dropping the rows with the missing values
df.dropna(subset=['Registration.Number'], inplace=True)

The rest with low percentages of missing values, I'll replace with either the mean/median for numerical data while for categorical data I'll either use the most common value or a placeholder value depending on the variable.

With the 'Schedule' column, I'll have to disregard it as a variable because the number of missing values is greater than actual values hence it's unviable.

#### Categorical Data

In [10]:
#checking col values to determine whether to replace with common value/placeholder value
df['Make'].value_counts()

Cessna            21959
Piper             11918
CESSNA             4920
Beech              4262
PIPER              2841
                  ...  
Loranger              1
Tankersley            1
Piper-aerostar        1
Patin                 1
Masko                 1
Name: Make, Length: 8223, dtype: int64

In [11]:
#filling the 'Make' values that are missing with a place holder value
df['Make'] = df['Make'].fillna('Unknown') 

In [12]:
#checking col values to determine whether to replace with common value/placeholder value
df['Model'].value_counts()

152           2362
172           1745
172N          1159
PA-28-140      928
150            826
              ... 
B757-232         1
Glasair RG       1
L 39C            1
L-1011-1         1
737 - 4D7        1
Name: Model, Length: 12089, dtype: int64

In [13]:
#filling the 'Model' values that are missing with a place holder value
df['Model'] = df['Model'].fillna('Uknown')

In [14]:
#checking col values to determine whether to replace with common value/placeholder value
df['Amateur.Built'].value_counts()

No     79101
Yes     8438
Name: Amateur.Built, dtype: int64

In [15]:
#filling the 'Amateur.Built' values that are missing with a place holder value
df['Amateur.Built'] = df['Amateur.Built'].fillna('Unknown')

In [16]:
#checking col values to determine whether to replace with common value/placeholder value
df['Engine.Type'].value_counts()

Reciprocating      69509
Turbo Shaft         3601
Turbo Prop          3387
Turbo Fan           2453
Unknown             1715
Turbo Jet            698
None                  19
Geared Turbofan       12
Electric              10
NONE                   2
LR                     2
UNK                    1
Hybrid Rocket          1
Name: Engine.Type, dtype: int64

In [17]:
#filling the 'Engine.Type' values that are missing with a place holder value
df['Engine.Type'] = df['Engine.Type'].fillna('Unknown')

In [18]:
#checking col values to determine whether to replace with common value/placeholder value
df['Purpose.of.flight'].value_counts()

Personal                     49271
Instructional                10548
Unknown                       6475
Aerial Application            4689
Business                      3965
Positioning                   1628
Other Work Use                1242
Ferry                          806
Aerial Observation             773
Public Aircraft                709
Executive/corporate            534
Flight Test                    401
Skydiving                      180
External Load                  120
Public Aircraft - Federal      103
Banner Tow                     101
Air Race show                   99
Public Aircraft - Local         74
Public Aircraft - State         63
Air Race/show                   59
Glider Tow                      53
Firefighting                    38
Air Drop                        10
ASHO                             6
PUBS                             4
PUBL                             1
Name: Purpose.of.flight, dtype: int64

In [19]:
#filling the 'Purpose.of.flight' values that are missing with a place holder value
df['Purpose.of.flight'] = df['Purpose.of.flight'].fillna('Unknown')

In [20]:
#checking col values to determine whether to replace with common value/placeholder value
df['Weather.Condition'].value_counts()

VMC    76709
IMC     5814
UNK      696
Unk      262
Name: Weather.Condition, dtype: int64

In [21]:
#filling the 'Weather.Condition' values that are missing with a common value
df['Weather.Condition'] = df['Weather.Condition'].fillna('VMC')  

In [22]:
#checking col values to determine whether to replace with common value/placeholder value
df['Aircraft.Category'].value_counts()

Airplane             27391
Helicopter            3394
Glider                 508
Balloon                231
Gyrocraft              173
Weight-Shift           161
Powered Parachute       91
Ultralight              30
Unknown                 14
WSFT                     9
Blimp                    4
Powered-Lift             4
UNK                      2
Rocket                   1
ULTR                     1
Name: Aircraft.Category, dtype: int64

In [23]:
#filling the 'Aircraft.Category' values that are missing with a place holder value
df['Aircraft.Category'] = df['Aircraft.Category'].fillna('Unknown')

In [24]:
#checking col values to determine whether to replace with common value/placeholder value
df['Broad.phase.of.flight'].value_counts()

Landing        15418
Takeoff        12478
Cruise         10269
Maneuvering     8144
Approach        6544
Climb           2034
Taxi            1952
Descent         1887
Go-around       1353
Standing         945
Unknown          548
Other            119
Name: Broad.phase.of.flight, dtype: int64

In [25]:
#filling the 'Broad.phase.of.flight' values that are missing with a place holder value
df['Broad.phase.of.flight'] = df['Broad.phase.of.flight'].fillna('Unknown')

#### Numerical Data

In [26]:
#checking the mean/mode for numeric columns to assess the best value to replace missing values
df[['Number.of.Engines','Total.Fatal.Injuries','Total.Serious.Injuries','Total.Minor.Injuries','Total.Uninjured']].describe()

Number.of.Engines  Total.Fatal.Injuries  Total.Serious.Injuries  \
count       82711.000000          76705.000000            75924.000000   
mean            1.146280              0.541399                0.265634   
std             0.445224              4.265354                1.392631   
min             0.000000              0.000000                0.000000   
25%             1.000000              0.000000                0.000000   
50%             1.000000              0.000000                0.000000   
75%             1.000000              0.000000                0.000000   
max             8.000000            295.000000              161.000000   

       Total.Minor.Injuries  Total.Uninjured  
count          76568.000000     82405.000000  
mean               0.339110         5.120029  
std                1.507999        27.200444  
min                0.000000         0.000000  
25%                0.000000         0.000000  
50%                0.000000         1.000000  
75%                0.000000         2.000000  
max              171.000000       699.000000

I've established that the `median` value is the best value because with the columns we have, they can only be represented by an integer.For example,we can't have 1.5 number of engines or injuries(people).

In [27]:
#filling missing values with the median in 'Number.of.Engines' column
df['Number.of.Engines'] = df['Number.of.Engines'].fillna(df['Number.of.Engines'].median())

In [28]:
#filling missing values with the median in 'Total.Fatal.Injuries' column
df['Total.Fatal.Injuries'] = df['Total.Fatal.Injuries'].fillna(df['Total.Fatal.Injuries'].median())

In [29]:
#filling missing values with the median in 'Total.Serious.Injuries' column
df['Total.Serious.Injuries'] = df['Total.Serious.Injuries'].fillna(df['Total.Serious.Injuries'].median())

In [30]:
#filling missing values with the median in 'Total.Minor.Injuries' column
df['Total.Minor.Injuries'] = df['Total.Minor.Injuries'].fillna(df['Total.Minor.Injuries'].median())

In [31]:
#filling missing values with the median in 'Total.Uninjured' column
df['Total.Uninjured'] = df['Total.Uninjured'].fillna(df['Total.Uninjured'].median())

**NB** The Schedule column still has missing values since we disregarded it.So let's remove it below.

In [32]:
working_cols.remove('Schedule')

In [33]:
df[working_cols].isna().sum()

Aircraft.Category         0
Registration.Number       0
Make                      0
Model                     0
Amateur.Built             0
Number.of.Engines         0
Engine.Type               0
Purpose.of.flight         0
Total.Fatal.Injuries      0
Total.Serious.Injuries    0
Total.Minor.Injuries      0
Total.Uninjured           0
Weather.Condition         0
Broad.phase.of.flight     0
dtype: int64